src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,167 @@
+        }
+        if (index == size) {
+            addLast(value);
+        } else {
+            addBefore(value, index);
+        }
     }
 
     @Override
     public boolean addAll(List<T> list) {
-        return false;
+        for (T element : list) {
+            add(element);
+        }
+        return true;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return getNode(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> tempNode = getNode(index);
+        T result = tempNode.value;
+        tempNode.value = value;
+        return result;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> tempNode = getNode(index);
+        T result = tempNode.value;
+        if (size == 1) {
+            first = null;
+            last = null;
+            size--;
+            return result;
+        }
+        if (tempNode.previous == null) {
+            first = tempNode.next;
+            tempNode.next.previous = null;
+            size--;
+            return result;
+        }
+        if (tempNode.next == null) {
+            last = tempNode.previous;
+            last.next = null;
+            size--;
+            return result;
+        }
+        tempNode.previous.next = tempNode.next;
+        tempNode.next.previous = tempNode.previous;
+        size--;
+        return result;
     }
 
     @Override
     public boolean remove(T t) {
+        for (int i = 0; i < size; i++) {
+            T check = get(i);
+            if (check == t || check != null && check.equals(t)) {
+                remove(i);
+                return true;
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void checkIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Not in range of the list!");
+        }
+    }
+
+    private void addBefore(T value, int index) {
+        Node<T> tempNode = getNode(index);
+        Node<T> newNode = new Node<>(value, tempNode.previous, tempNode);
+        if (tempNode.previous == null) {
+            tempNode.next.previous = newNode;
+            first = newNode;
+        } else {
+            tempNode.previous.next = newNode;
+            tempNode.previous = newNode;
+        }
+        size++;
+    }
+
+    private void addLast(T value) {
+        Node<T> newNode = new Node<>(value, last, null);
+        if (first == null) {
+            newNode.previous = null;
+            first = newNode;
+            last = newNode;
+        } else {
+            last.next = newNode;
+            last = newNode;
```

Comment:
this line of code repeats itself, let's move it outside if else

Reply:
Hm, really changed it.